### PR TITLE
#371: [docs] use fixed doxygen version, add C++17 dependency, fix markup issues

### LINKED ISSUE #371: [docs] add C++17 dependency, fix markup issues
> `std::optional` is documented as being added in C++17, while [Magistrate documentation](https://darma-tasking.github.io/checkpoint_docs/html/index.html) says it's a C++14 library.
> 
> Fortunately, it doesn't look like `std::optional` is actually *used* anywhere in here, so this inclusion should just be dropped.
> 
> Also, maybe that line of the documentation is out of date, in which case, please open an issue. Additionally, I see two more issues of documentation markup not being rendered properly, which may represent additional issues(s):
> * `\checkpoint`
> * `@m_class{m-note m-dim}`
> 

_Originally posted by @PhilMiller in https://github.com/DARMA-tasking/magistrate/pull/345#discussion_r1774256473_

**TODO**:
- `magistrate` requires C++17, update the docs to reflect that.
- Take a look at the rendering issues mentioned by Phil and fix them or open a separate issue.

### PR DESCRIPTION
**Pull request by cz4rs**
fixes #371 

As the title says, plus a bunch of `checkpoint` -> `magistrate` renaming changes.

Successful build from this branch: https://github.com/DARMA-tasking/magistrate/actions/runs/11805143657/job/32886943659?pr=375.

### CODE REVIEW

**Comment by cz4rs** [comment]
Before:
image

After:
image

File: docs/md/mainpage.md [hunk 1]
```diff
@@ -1,22 +1,18 @@
-\mainpage Introduction
+\mainpage DARMA/magistrate => serialization and checkpointing library
 
-\tableofcontents
```
Comment (cz4rs) [inline docs/md/mainpage.md:3]: Let's drop TOC, this page is really short.

File: docs/md/mainpage.md [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
 it to be recursively traversed and transformed by the library. For instance, if
 the user wants to send over the network, the traversal will create a byte buffer
-that contains all the data for a message. `DARMA/checkpoint` is a general
+that contains all the data for a message. `DARMA/magistrate` is a general
 library that can also be used for other purposes such as I/O---creating
 snapshots of objects to disk.
 
 \section License
 
-@m_class{m-note m-dim}
```
Comment (cz4rs) [inline docs/md/mainpage.md]: It's parsed correctly now, but still doesn't have the desired dim effect. I might give it one more try tomorrow :shrug:

File: docs/shared/checkpoint_learn_serialize.md [hunk 3]
```diff
@@ -83,10 +83,10 @@ serialization of the class state.
 To serialize polymorphic class hierarchies, one must write serializers for each
 class in the hierarchy. Then, the user should either insert macros
 `checkpoint_virtual_serialize_root()` and
-`checkpoint_virtual_serialize_derived_from(T)` to inform *checkpoint* of the
+`checkpoint_virtual_serialize_derived_from(T)` to inform *magistrate* of the
```
Comment (cz4rs) [inline docs/shared/checkpoint_learn_serialize.md:86]: Note that `checkpoint_virtual_serialize` macros were never renamed to `magistrate_virtual_serialize`.

File: docs/md/mainpage.md (hunk 2 above)
Comment (cz4rs) [reply docs/md/mainpage.md]: Fixed.

**Review by lifflander — APPROVED** [review]

File: docs/shared/checkpoint_learn_serialize.md (hunk 3 above)
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md:86]: Can we include this change in the PR?

**Review by cwschilly — APPROVED** [review]
LGTM

**Pull request merged**